File: Project/Backend/src/ArtAuction.API/Controllers/WatchListController.cs
using ArtAuction.Application.DTOs.WatchList;
using ArtAuction.Application.Interfaces.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace ArtAuction.API.Controllers;

[ApiController]
[Route("[controller]")]
[Authorize(Roles = "Buyer")]
public class WatchListController : ControllerBase
{
    // Attributes
    private readonly IWatchListServices _watchListServices;

    // Constructor
    public WatchListController(IWatchListServices watchListServices)
    {
        this._watchListServices = watchListServices;
    }

    // Get buyer watch list
    [HttpGet("GetWatchListForBuyer")]
    public async Task<IActionResult> GetWatchListForBuyer(int buyerId)
    {
        // Return watch list
        return Ok(await _watchListServices.GetWatchListForBuyer(buyerId));
    }

    // Create watch relation API
    [HttpPost("CreateWatchList")]
    public async Task<IActionResult> CreateWatchList([FromBody] WatchListCreationDto watchListCreationDto)
    {
        // Check model state
        if (!ModelState.IsValid)
            return BadRequest(ModelState);

        // Check stats
        var result = await _watchListServices.CreateWatchList(watchListCreationDto);
        if (!result)
            return BadRequest("Could not create watchlist relation. Check IDs or duplicate relation.");

        // Return success stats
        return Ok("Watchlist relation created successfully.");
    }

    // Delete watch relation API
    [HttpDelete("DeleteWatchList")]
    public async Task<IActionResult> DeleteWatchList(int buyerId, int artworkPostId)
    {
        // Check stats
        var result = await _watchListServices.DeleteWatchList(buyerId, artworkPostId);
        if (!result)
            return NotFound("Watchlist relation was not found.");

        // Return success stats
        return Ok("Watchlist relation deleted successfully.");
    }
}

File: ArtworkPostServices/src/main/java/com/artauction/artworkpostservices/Controllers/WatchListController.java
package com.artauction.artworkpostservices.Controllers;

import com.artauction.artworkpostservices.Aops.Roles.RequiresRole;
import com.artauction.artworkpostservices.Dtos.WatchList.WatchListCreationDto;
import com.artauction.artworkpostservices.Services.Interfaces.WatchListServices;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/WatchList")
@RequiresRole("Buyer")
@RequiredArgsConstructor
public class WatchListController {
    // Attributes
    private final WatchListServices watchListServices;

    // Get buyer watch list
    @GetMapping("/GetWatchListForBuyer")
    public ResponseEntity<?> getWatchListForBuyer(@RequestParam Integer buyerId) {
        return ResponseEntity.ok(watchListServices.getWatchListForBuyer(buyerId));
    }

    // Create watch artwork post API
    @PostMapping("/CreateWatchList")
    public ResponseEntity<?> createWatchList(@Valid @RequestBody WatchListCreationDto watchListCreationDto) {
        // Create and check result
        boolean result = watchListServices.createWatchList(watchListCreationDto);
        if (!result)
            return ResponseEntity.badRequest()
                    .body("Could not create watchlist relation. Check IDs or duplicate relation.");

        // Return success stats
        return ResponseEntity.ok("Watchlist relation created successfully.");
    }

    // Delete watch artwork post API
    @DeleteMapping("/DeleteWatchList")
    public ResponseEntity<?> deleteWatchList(
            @RequestParam Integer buyerId,
            @RequestParam Integer artworkPostId
    ) {
        // Delete and check result
        boolean result = watchListServices.deleteWatchList(buyerId, artworkPostId);
        if (!result)
            return ResponseEntity.status(404).body("Watchlist relation was not found.");

        // Return success stats
        return ResponseEntity.ok("Watchlist relation deleted successfully.");
    }
}
